source/plugins/go/input/lib/kubernetesapiclient.go

```diff
@@ -0,0 +1,332 @@
+	LogPath                     string
+	logger                      *log.Logger
+	TokenExpiry                 int64
+)
+
+func init() {
+	// Define log path
+	isTestEnv := os.Getenv("ISTEST") == "true"
+	osType := os.Getenv("OS_TYPE")
+	if osType != "" && osType == "windows" {
+		LogPath = WindowsLogPath + "kubernetes_client_log.txt"
+	} else {
+		LogPath = LinuxLogPath + "kubernetes_client_log.txt"
+	}
+
+	if isTestEnv {
+		LogPath = "./kubernetes_client_log.txt"
+	}
+
+	TokenExpiry = time.Now().Unix()
+
+	// Define logger
+	file, err := os.OpenFile(LogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+	if err != nil {
+		log.Fatalln("Failed to open log file :", err)
+	}
+	logger = log.New(file, "", log.LstdFlags)
+}
+
+func getKubeResourceInfo(resource string, api_group *string) (*http.Response, error) {
+	logger.Println("Getting Kube resource: ", resource)
+
+	resourceUri, err := getResourceUri(resource, api_group)
+	if err != nil {
+		logger.Println("getResourceUri failed: ", err)
+		return nil, err
+	}
+
+	// Load CA cert
+	caCert, err := ioutil.ReadFile(CaFile)
+	if err != nil {
+		logger.Println("Failed to read ca.crt: ", err)
+		return nil, err
+	}
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		RootCAs: caCertPool,
+	}
+
+	// Setup HTTPS client
+	tlsConfig.BuildNameToCertificate()
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+	client := &http.Client{Transport: transport, Timeout: time.Second * 40}
+
+	req, _ := http.NewRequest("GET", resourceUri, nil)
+	req.Header.Add("Authorization", "Bearer "+GetTokenStr())
+
+	resp, err := client.Do(req)
+	if err != nil {
+		logger.Println("Failed to make request to Kube API: ", err)
+		return nil, err
+	}
+
+	logger.Println("KubernetesAPIClient::getKubeResourceInfo : Got response of ", resp.StatusCode)
+
+	return resp, nil
+}
+
+func getResourceUri(resource string, api_group *string) (string, error) {
+	serviceHost, serviceHostExist := os.LookupEnv("KUBERNETES_SERVICE_HOST")
+	servicePort, servicePortExist := os.LookupEnv("KUBERNETES_PORT_443_TCP_PORT")
+
+	if serviceHostExist && servicePortExist {
+		switch {
+		case api_group == nil:
+		    return "https://" + serviceHost + ":" + servicePort + "/api/" + ApiVersion + "/" + resource, nil
+		case *api_group == ApiGroupApps:
+		    return "https://" + serviceHost + ":" + servicePort + "/apis/apps/" + ApiVersionApps + "/" + resource, nil
+		case *api_group == ApiGroupHPA:
+		    return "https://" + serviceHost + ":" + servicePort + "/apis/" + ApiGroupHPA + "/" + ApiVersionHPA + "/" + resource, nil
+		default:
+		    return "", fmt.Errorf("unsupported API group: %s", *api_group)
+		}
+	} else {
+		logger.Println("Kubernetes environment variable not set KUBERNETES_SERVICE_HOST: ", serviceHost, " KUBERNETES_PORT_443_TCP_PORT: ", servicePort, ". Unable to form resourceUri")
+		return "", errors.New("environment variables not set")
+	}
+}
+
+func GetTokenStr() string {
```

Comment:
This is used only for k8s APIs. why do we need this perf and container inventory which should use cAdvisor?

Reply:
mostly for GetClusterName functionality